src/main/java/vazkii/patchouli/client/book/text/SpanState.java

```diff
@@ -83,4 +101,37 @@ public void reset() {
 	public Style peekStyle() {
 		return styleStack.getFirst();
 	}
+
+        // Represents the transformations on a single stack frame.
+        private class SpanPartialState {
+                // This is null iff this state represents the base state (i.e. nothing to merge)
+                private final Style mergeStyle;
+                // List of all the transformations that are applied to this frame.
+                private List<Function<Style, Style>> transformations = null;
+
+                public SpanPartialState(Style style) {
+                        this.mergeStyle = style;
+                }
+
+                public void addModification(Function<Style, Style> f) {
+                        if (transformations == null) {
+                                // We use a linked list because this list is likely to be very small
+                                // in order to skimp on space.
+                                transformations = new LinkedList<>();
```

Comment:
does this matter that much? this is all thrown away after layout, right? same with the lazy init thing

Reply:
honestly, unless the user makes a _huge_ layout change, it shouldn't matter. i'm cheap anyway :wacko: